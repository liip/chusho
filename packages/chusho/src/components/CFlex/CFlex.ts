import {
  defineComponent,
  createElement,
  InjectionKey,
  provide,
} from '@vue/composition-api';
import { VNodeData } from 'vue/types/umd';
import { mergeData } from 'vue-functional-data-merge';
import { ContainerItemClass } from '@/types';
import { warn } from '@/utils/debug';

export const FlexSymbol: InjectionKey<object> = Symbol();

export interface UseFlex {
  gapX?: ContainerItemClass;
  gapY?: ContainerItemClass;
}

interface FlexProps {
  gap?: string | number | Array<string>;
}

export default defineComponent<FlexProps>({
  name: 'CFlex',

  props: {
    /**
     * One of the gaps defined in the config (flex.gaps) to apply to the element.
     *
     * A value of type string apply the same gap on both axis. Use an array of type `[x, y]` to have a different gap between columns and rows.
     */
    gap: {
      type: [String, Number, Array],
      default: null,
    },
  },

  setup(props, { attrs, slots, parent }) {
    const flexConfig = parent!.$chusho.options.components?.flex;
    const gapXKey = props.gap instanceof Array ? props.gap[0] : props.gap;
    const gapYKey = props.gap instanceof Array ? props.gap[1] : props.gap;
    const gapX = gapXKey ? flexConfig?.gaps?.x?.[gapXKey] : undefined;
    const gapY = gapYKey ? flexConfig?.gaps?.y?.[gapYKey] : undefined;

    const api: UseFlex = {
      gapX,
      gapY,
    };
    provide(FlexSymbol, api);

    if (props.gap && !gapX) {
      warn(
        `Cannot find a Flex gap named "${gapXKey}" under "flex.gaps.x" defined in the config, you must define it before referencing it.`
      );
    }
    if (props.gap && !gapY) {
      warn(
        `Cannot find a Flex gap named "${gapYKey}" under "flex.gaps.y" defined in the config, you must define it before referencing it.`
      );
    }

    return () => {
      const componentData: VNodeData = mergeData(
        {
          ...attrs,
        },
        {
          class: [
            flexConfig?.containerClass,
            gapX?.containerClass,
            gapY?.containerClass,
          ],
        }
      );

      return createElement(
        'div',
        componentData,
        slots.default && slots.default()
      );
    };
  },
});
