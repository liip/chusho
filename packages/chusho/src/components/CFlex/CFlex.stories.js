import CFlex from './CFlex';
import CFlexItem from './CFlexItem';

export default {
  title: 'Components|Flex',
  component: CFlex,
  subcomponents: { CFlexItem },
  parameters: {
    componentSubtitle: 'Layout content in rows or columns responsively.',
    options: {
      componentConfig: [
        {
          name: 'containerClass',
          type: { summary: 'string' },
          description: 'Classes to apply to all Flex containers.',
        },
        {
          name: 'itemClass',
          type: { summary: 'string' },
          description: 'Classes to apply to all Flex items.',
        },
        {
          name: 'widths',
          type: { summary: 'object' },
          description:
            "List of widths/classes pair CFlexItem will apply based on its `width` prop.\n\nFor example: `{ '1/2': 'w-1/2', '1/3': 'w-1/3', fallback: 'w-full' }`\n\nNotice the special `fallback` key that will be apply if no `default` value is given ",
        },
        {
          name: 'gaps',
          type: { summary: 'object' },
          description:
            'List of available gaps.\n\nFor example: `{ "1": { "containerClass": "-mt-1", "itemClass": "mt-1" } }`',
        },
      ],
    },
  },
};

export const GridResponsive = () => ({
  components: { CFlex, CFlexItem },
  template: `<CFlex gap="2">
    <CFlexItem width="1/3">
      <div class="bg-gray-200 p-2">
        1
      </div>
    </CFlexItem>
    <CFlexItem width="1/3">
      <div class="bg-gray-200 p-2">
        2
      </div>
    </CFlexItem>
    <CFlexItem width="1/3">
      <div class="bg-gray-200 p-2">
        3
      </div>
    </CFlexItem>
    <CFlexItem :width="{ md: '2/3' }">
      <div class="bg-gray-200 p-2">
        4
      </div>
    </CFlexItem>
    <CFlexItem :width="{ md: '1/3' }">
      <div class="bg-gray-200 p-2">
        5
      </div>
    </CFlexItem>
    <CFlexItem :width="{ sm: '1/3', lg: '1/2' }">
      <div class="bg-gray-200 p-2">
        6
      </div>
    </CFlexItem>
    <CFlexItem :width="{ sm: '1/3', lg: '1/2' }">
      <div class="bg-gray-200 p-2">
        7
      </div>
    </CFlexItem>
  </CFlex>`,
});

export const DifferentGapByAxis = () => ({
  components: { CFlex, CFlexItem },
  template: `<CFlex :gap="[3, 6]">
    <CFlexItem width="1/3">
      <div class="bg-gray-200 p-2">
        1
      </div>
    </CFlexItem>
    <CFlexItem width="1/3">
      <div class="bg-gray-200 p-2">
        2
      </div>
    </CFlexItem>
    <CFlexItem width="1/3">
      <div class="bg-gray-200 p-2">
        3
      </div>
    </CFlexItem>
    <CFlexItem :width="{ md: '2/3', lg: '1/2' }">
      <div class="bg-gray-200 p-2">
        4
      </div>
    </CFlexItem>
    <CFlexItem :width="{ md: '1/3', lg: '1/2' }">
      <div class="bg-gray-200 p-2">
        5
      </div>
    </CFlexItem>
  </CFlex>`,
});

export const Inline = () => ({
  components: { CFlex, CFlexItem },
  template: `<CFlex gap="2">
    <CFlexItem><div class="bg-gray-200 p-2">Alice Blue</div></CFlexItem>
    <CFlexItem><div class="bg-gray-200 p-2">Antique White</div></CFlexItem>
    <CFlexItem><div class="bg-gray-200 p-2">Aqua</div></CFlexItem>
    <CFlexItem><div class="bg-gray-200 p-2">Aquamarine</div></CFlexItem>
    <CFlexItem><div class="bg-gray-200 p-2">Azure</div></CFlexItem>
    <CFlexItem><div class="bg-gray-200 p-2">Beige</div></CFlexItem>
    <CFlexItem><div class="bg-gray-200 p-2">Bisque</div></CFlexItem>
    <CFlexItem><div class="bg-gray-200 p-2">Black</div></CFlexItem>
    <CFlexItem><div class="bg-gray-200 p-2">Blanched Almond</div></CFlexItem>
    <CFlexItem><div class="bg-gray-200 p-2">Blue</div></CFlexItem>
    <CFlexItem><div class="bg-gray-200 p-2">Blue Violet</div></CFlexItem>
    <CFlexItem><div class="bg-gray-200 p-2">Brown</div></CFlexItem>
    <CFlexItem><div class="bg-gray-200 p-2">Burly Wood</div></CFlexItem>
    <CFlexItem><div class="bg-gray-200 p-2">Cadet Blue</div></CFlexItem>
    <CFlexItem><div class="bg-gray-200 p-2">Chartreuse</div></CFlexItem>
    <CFlexItem><div class="bg-gray-200 p-2">Chocolate</div></CFlexItem>
    <CFlexItem><div class="bg-gray-200 p-2">Coral</div></CFlexItem>
    <CFlexItem><div class="bg-gray-200 p-2">Cornflower Blue</div></CFlexItem>
    <CFlexItem><div class="bg-gray-200 p-2">Cornsilk</div></CFlexItem>
    <CFlexItem><div class="bg-gray-200 p-2">Crimson</div></CFlexItem>
    <CFlexItem><div class="bg-gray-200 p-2">Cyan</div></CFlexItem>
    <CFlexItem><div class="bg-gray-200 p-2">Dark Blue</div></CFlexItem>
    <CFlexItem><div class="bg-gray-200 p-2">Dark Cyan</div></CFlexItem>
    <CFlexItem><div class="bg-gray-200 p-2">Dark GoldenRod</div></CFlexItem>
    <CFlexItem><div class="bg-gray-200 p-2">Dark Gray</div></CFlexItem>
    <CFlexItem><div class="bg-gray-200 p-2">Dark Green</div></CFlexItem>
  </CFlex>`,
});

export const Stack = () => ({
  components: { CFlex, CFlexItem },
  template: `<CFlex gap="3" class="flex-col">
    <CFlexItem width="full">
      <div class="p-2 bg-gray-200">
        1
      </div>
    </CFlexItem>
    <CFlexItem width="full">
      <div class="p-2 bg-gray-200">
        2
      </div>
    </CFlexItem>
    <CFlexItem width="full">
      <div class="p-2 bg-gray-200">
        3
      </div>
    </CFlexItem>
  </CFlex>`,
});
