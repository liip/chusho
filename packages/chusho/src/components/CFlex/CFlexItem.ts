import { defineComponent, createElement, inject } from '@vue/composition-api';
import { warn } from '@/utils/debug';
import { mergeData } from 'vue-functional-data-merge';
import { VNodeData } from 'vue/types/umd';
import { FlexSymbol, UseFlex } from './CFlex';
import { isPlainObject } from '@/utils/objects';

interface FlexItemProps {
  width: string | Dictionary<string>;
}

export default defineComponent<FlexItemProps>({
  name: 'CFlexItem',

  props: {
    /**
     * One or more widths defined in the config (flex.widths) to apply to the element.
     *
     * Single default value: `width="1/2"`
     *
     * Multiple values: `:width="{ default: '1/2', sm: '1/3', md: '1/4' }"`
     *
     * Except the `default` key which acts like the single value, all others (`sm` & `md` here) are totally up to you. They should probably match your own breakpoints naming though.
     *
     * You can change how responsive classes are generated by defining a custom `responsiveClassGenerator` function in the config.
     */
    width: {
      type: [String, Object],
    },
  },

  setup(props, { slots, attrs, parent }) {
    const chushoOptions = parent!.$chusho.options;
    const flexConfig = chushoOptions.components?.flex;
    const flex = inject(FlexSymbol) as UseFlex;
    const widthClasses: Array<string> = [];
    let widths: Dictionary<string>;

    if (props.width) {
      if (isPlainObject(props.width)) {
        widths = props.width;
      } else {
        widths = {
          default: props.width,
        };
      }

      const breakpoints = Object.keys(widths);

      breakpoints.forEach(breakpoint => {
        const value = widths[breakpoint];
        const width = flexConfig?.widths?.[value];

        if (width) {
          if (breakpoint === 'default') {
            widthClasses.push(width);
            return;
          }
          widthClasses.push(
            chushoOptions.responsiveClassGenerator(breakpoint, width)
          );
        } else {
          warn(
            `Cannot find a FlexItem width named "${value}" defined under "flex.widths" in the config, you must define it before referencing it.`
          );
        }
      });

      const hasDefault = breakpoints.includes('default');
      if (!hasDefault && flexConfig?.widths?.fallback) {
        widthClasses.push(flexConfig.widths.fallback);
      }
    }

    return () => {
      const componentData: VNodeData = mergeData(
        {
          ...attrs,
        },
        {
          class: [
            ...widthClasses,
            flexConfig?.itemClass,
            flex.gapX?.itemClass,
            flex.gapY?.itemClass,
          ],
        }
      );

      return createElement(
        'div',
        componentData,
        slots.default && slots.default()
      );
    };
  },
});
