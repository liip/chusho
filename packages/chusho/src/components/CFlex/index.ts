import CFlex from './CFlex';
import CFlexItem from './CFlexItem';

export { CFlex, CFlexItem };
export default CFlex;
