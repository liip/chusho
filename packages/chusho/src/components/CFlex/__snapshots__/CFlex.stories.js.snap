// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots Components|Flex Default 1`] = `
<div
  class="flex flex-wrap -ml-2 -mt-2"
>
  <div
    class="w-1/3 pl-2 pt-2"
  >
    <div
      class="bg-gray-200 p-2"
    >

        1

    </div>
  </div>

  <div
    class="w-1/3 pl-2 pt-2"
  >
    <div
      class="bg-gray-200 p-2"
    >

        2

    </div>
  </div>

  <div
    class="w-1/3 pl-2 pt-2"
  >
    <div
      class="bg-gray-200 p-2"
    >

        3

    </div>
  </div>

  <div
    class="md:w-2/3 w-full pl-2 pt-2"
  >
    <div
      class="bg-gray-200 p-2"
    >

        4

    </div>
  </div>

  <div
    class="md:w-1/3 w-full pl-2 pt-2"
  >
    <div
      class="bg-gray-200 p-2"
    >

        5

    </div>
  </div>

  <div
    class="sm:w-1/3 lg:w-1/2 w-full pl-2 pt-2"
  >
    <div
      class="bg-gray-200 p-2"
    >

        6

    </div>
  </div>

  <div
    class="sm:w-1/3 lg:w-1/2 w-full pl-2 pt-2"
  >
    <div
      class="bg-gray-200 p-2"
    >

        7

    </div>
  </div>
</div>
`;

exports[`Storyshots Components|Flex Different Gap By Axis 1`] = `
<div
  class="flex flex-wrap -ml-3 -mt-6"
>
  <div
    class="w-1/3 pl-3 pt-6"
  >
    <div
      class="bg-gray-200 p-2"
    >

        1

    </div>
  </div>

  <div
    class="w-1/3 pl-3 pt-6"
  >
    <div
      class="bg-gray-200 p-2"
    >

        2

    </div>
  </div>

  <div
    class="w-1/3 pl-3 pt-6"
  >
    <div
      class="bg-gray-200 p-2"
    >

        3

    </div>
  </div>

  <div
    class="md:w-2/3 lg:w-1/2 w-full pl-3 pt-6"
  >
    <div
      class="bg-gray-200 p-2"
    >

        4

    </div>
  </div>

  <div
    class="md:w-1/3 lg:w-1/2 w-full pl-3 pt-6"
  >
    <div
      class="bg-gray-200 p-2"
    >

        5

    </div>
  </div>
</div>
`;

exports[`Storyshots Components|Flex No Gap 1`] = `
<div
  class="flex flex-wrap"
>
  <div
    class="w-1/2 lg:w-1/4"
  >
    <div
      class="bg-gray-200 p-2"
    >

        1

    </div>
  </div>

  <div
    class="w-1/2 lg:w-1/4"
  >
    <div
      class="bg-gray-200 p-2"
    >

        2

    </div>
  </div>

  <div
    class="w-1/2 lg:w-1/4"
  >
    <div
      class="bg-gray-200 p-2"
    >

        3

    </div>
  </div>

  <div
    class="w-1/2 lg:w-1/4"
  >
    <div
      class="bg-gray-200 p-2"
    >

        4

    </div>
  </div>

  <div
    class="w-1/2 lg:w-1/4"
  >
    <div
      class="bg-gray-200 p-2"
    >

        5

    </div>
  </div>

  <div
    class="w-1/2 lg:w-1/4"
  >
    <div
      class="bg-gray-200 p-2"
    >

        6

    </div>
  </div>
</div>
`;
